security/pkg/pki/ca/controller/secret.go

```diff
@@ -72,13 +75,18 @@ type SecretController struct {
 }
 
 // NewSecretController returns a pointer to a newly constructed SecretController instance.
-func NewSecretController(ca ca.CertificateAuthority, certTTL time.Duration, core corev1.CoreV1Interface,
-	namespace string) *SecretController {
+func NewSecretController(ca ca.CertificateAuthority, certTTL time.Duration, gracePeriodRatio float32, core corev1.CoreV1Interface,
+	namespace string) (*SecretController, error) {
+
+	if gracePeriodRatio < 0 || gracePeriodRatio > 1 {
```

Comment:
Should we also have some validation of the combination of TTL and grace period? Some ideas:

- Should we define a minimum allowed TTL in order to guarantee that we always have some reasonable grace period?
- Should we define a min/max for `gracePeriodRatio` as well? Would we ever want it to be exactly 0 or 1? Perhaps we just need to apply some min/max for the calculated `gracePeriod` (based on the ttl)?

I would also suggest that if a bogus configuration forces us to use our min/max values, that we log a warning to the user.

Reply:
What you said make sense. I have added the warning for values outside of the recommended window (for now it's [0.5, 0.8]). For min grace period, I haven't decided whether we want to add that, my major concern is this may confuse our users. So I added a TODO.

Reply:
@myidpt my concern is that a bad configuration can put us into a non-functional state. I don't think logging a warning is strong enough here.

WRT confusing the users, what exactly do you see as the concern?  The parameters themselves perhaps?  Perhaps configuring a ratio is too complex, leaving the user to do the math? If so, we can reconsider using `ttl` and `gracePeriod` instead.

Either way, we'll want to put hard limits here so the user can't shoot themselves in the foot.

Reply:
IMHO, since there is a default value for the gracePeriodRatio (0.5), most of our users won't touch it. For the advanced users that want to change the gracePeriodRatio, they will be cautious enough to make sure the gracePeriod is reasonable.

I think having gracePeriod is less preferable because in the future we may have certs with different TTLs. Applying the same gracePeriod on them may not work, especially for pretty short-lived certs.